lib/executor/vu_handle.go

```diff
@@ -77,52 +87,51 @@ func newStoppedVUHandle(
 
 	// TODO maybe move the repeating parts in a function?
 	vh.returnVU = func(v lib.InitializedVU) {
-		// Don't return the initialized VU back
 		vh.mutex.Lock()
-		select {
-		case <-vh.parentCtx.Done():
-			// we are done just ruturn the VU
+		// the returnVU could be called after those have reinitialize, don't nil them if that is the
+		// case, maybe just don't nil them
+		if vh.state != starting && vh.state != running {
 			vh.initVU = nil
 			vh.activeVU = nil
-			atomic.StoreInt32(&vh.change, 1)
-			vh.mutex.Unlock()
-			returnVU(v)
-		default:
-			select {
-			case <-vh.canStartIter:
-				// we can continue with itearting - lets not return the vu
-				vh.activateVU() // we still need to reactivate it to get the new context and cancel
-				atomic.StoreInt32(&vh.change, 1)
-				vh.mutex.Unlock()
-			default: // we actually have to return the vu
-				vh.initVU = nil
-				vh.activeVU = nil
-				atomic.StoreInt32(&vh.change, 1)
-				vh.mutex.Unlock()
-				returnVU(v)
-			}
 		}
+		vh.mutex.Unlock()
+		returnVU(v)
 	}
 
 	return vh
 }
 
+// start is one of the more complicated ones as it can race with toGracefulStop and has special
+// cases
+// state changes
+// toGracefulStop => running, we have raced with toGracefulStop if it was finished the state
+// would've been stoppped
+// starting, running => nothing, we have done what needs to be done
+// stopped, toHardStop => reactivate everything, it is either stopped or hardStop has uninitialized
+// some stuff and we need to reinitialize them
```

Comment:
Would you mind creating a [state-transition table](https://en.wikipedia.org/wiki/State-transition_table) (or a state diagram in something like graphviz/DOT if you're inspired :smile:) for this? That way one can reference it for all methods, and it's a bit easier to read.

Reply:
I had a drawing (on paper with pen), which got outdated ... way too long ago, and I intented on having it as a comment in ascii art, but a state-transition table seems better

Reply:
fixed ?

Reply:
Yeah, thanks, that's much easier to understand now. I think the transitions make sense, but will have to dig a bit deeper.